### DISCUSSION TITLE
How to build, start and use the web server and web app for configuration?

### DISCUSSION THREAD

**Discussion by maehw** [xtech/self-o-mat#14] [Q&A] [answered]
Hi @ClemensElflein,

maybe you could provide some details here. I think the web server and web app are the least documented (but probably quite powerful) feature of self-o-mat. This repo's [current README.md](https://github.com/xtech/self-o-mat/blob/f54443a7812b7ede709f2addcf5f76b2f5334100/README.md) only mentions it as a feature but not how the web server is built, started and used to configure the self-o-mat photo booth. I'd like to shed some light on this.

> App Inclusive - update camera settings on the go. Just connect your smartphone and change settings live!

However, I've collected some hints here:

* [served](https://github.com/meltwater/served/tree/8ef95d72692ef1f392fc0a44a6dad4235386ac9d) is used as a submodule
  >Served is a C++ library for building high performance RESTful web servers. Served builds upon Boost.ASIO to provide a simple API for developers to create HTTP services in C++.)
* https://www.self-o-mat.de/ mentions that you can perform settings with an app or web interface
  > Einstellung per App
Ändere alle Kameraeinstellungen und Fotoboxeinstellungen bequem direkt am Smartphone. Die App ist per Webinterface mit jedem Gerät aufrufbar. Alternativ gibt es eine native Version für iOS Geräte.
* The assembly manual mentions the app in the printer installation section and also provides some screenshots.
  >Open self-o-mat app:
  >WLAN: self-o-mat
  >Passwort: self-o-mat
  >URL: http://192.168.4.1:9080/app
  >You find a printer installation link in the photobooth-settings (CUPS Printer Setup).
  >Alternatively you can open the following URL in your browser: http://192.168.4.1:631/

  So it seems that the Raspberry Pi provides a WiFi access point - and a web server.

I've the following questions:

* How is the WiFi access point configured? Those scripts and configuration files are probably not (yet) part of the repository. I think they could be extracted from the SD card images you provide for download. Would be cool to have them here to re-build such a system manually.
* How do you build the web server? Is it included in the build process via `cd self-o-mat && mkdir build && cd build && cmake .. && make`? There are other shell scripts that might relate: `build_web.sh` and `deploy_app.sh`
* How do you run the web server? Is this automatically done by calling `./start.sh` - or is it done independently?
* Are web server and/or application automatically started on boot? If so, how? There will be additional scripts and/or configuration file (that will be part of the SD card image and could be extracted and tracked in this repo).
* How do you access the web server for configuration? In #9 you've mentioned that the native app is no longer maintained - but that using a browser should do the trick. So the short answer is probably: know the IP of your self-o-mat and open `http://<IP>:9080/app` in your browser when you're connected to the WiFi.
* Not really a question but a hint (for later documentation): there is one web server for self-o-mat configuration (port: ``) and another web server for CUPS configuration.
* What is the interface between the web server and the standalone photo booth application? It's probably JSON files, but I am not sure (there may be sockets other other ways, too).
* When are settings changes made via the web server applied to the standalone photo booth application? Immediately, after restart of the standalone application or does the web app even restart the photo booth standalone application, after reboot?

**Comment by ClemensElflein** [comment]
>       So it seems that the Raspberry Pi provides a WiFi access point - and a web server.
That is correct!

> I've the following questions:
> 
>     * How is the WiFi access point configured? Those scripts and configuration files are probably not (yet) part of the repository. I think they could be extracted from the SD card images you provide for download. Would be cool to have them here to re-build such a system manually.

That is a good idea. To be honest, I have no idea anymore. It has been > 2 years since I created this config.
IIRC there is a hostapd config file and an NGINX config file. The hostapd config file sets up the access point and the NGINX config file redirects to the web app if it is present and to a fallback static page on timeout.
But there is some additional magic going on for the hotspot detection feature - you probably know it as the "you are connected to a hotspot, click here to login"-thing.
I'd have to reverse engineer the config files in the image myself.
I have created issue #16 for this. I'll get to it if I have some free time or maybe someone else wants to do this?

> 
>     * How do you build the web server? Is it included in the build process via `cd self-o-mat && mkdir build && cd build && cmake .. && make`? There are other shell scripts that might relate: `build_web.sh` and `deploy_app.sh`

The web server is included in the self-o-mat software and starts automatically.

- build_web.sh: Builds the web app from source, you don't need to do this as the app is already built in the 'web' folder.
- build_app.sh: Is used to create selfomat.update files. These files contain all self-o-mat binaries and dependencies and are signed for use with the updater running on official self-o-mat boxes. You also don't need this file.

> 
>     * How do you run the web server? Is this automatically done by calling `./start.sh` - or is it done independently?

Yes, it starts with the software. The web server is accessible on port 9080 on all interfaces.
Just open a browser to http://localhost:9080

> 
>     * Are web server and/or application automatically started on boot? If so, how? There will be additional scripts and/or configuration file (that will be part of the SD card image and could be extracted and tracked in this repo).'

We have built a bootloader app [up-2-date](https://github.com/xtech/up-2-date) which runs on boot. It shows a nice gui, reads update files from SD card, verifies signatures and starts the photo booth app. I think for a DIY solution this might be overkill. Just create a systemd service.

> 
>     * How do you access the web server for configuration? In [How to configure everything? #9](https://github.com/xtech/self-o-mat/discussions/9) you've mentioned that the native app is no longer maintained - but that using a browser should do the trick. So the short answer is probably: know the IP of your self-o-mat and open `http://<IP>:9080/app` in your browser when you're connected to the WiFi.

This is correct.

> 
>     * Not really a question but a hint (for later documentation): there is one web server for self-o-mat configuration (port: ``) and another web server for CUPS configuration.
This is also correct. Although the self-o-mat app links to the CUPS page so you don't need to remember both.

> 
>     * What is the interface between the web server and the standalone photo booth application? It's probably JSON files, but I am not sure (there may be sockets other other ways, too).
It's Protobuf. The message definitions can be found here: [https://github.com/xtech/self-o-mat/blob/master/api.proto](https://github.com/xtech/self-o-mat/blob/master/api.proto)
> 
>     * When are settings changes made via the web server applied to the standalone photo booth application? Immediately, after restart of the standalone application or does the web app even restart the photo booth standalone application, after reboot?

Changes are applied instantly since the web server directly calls C++ functions. All settings are stored on photo booth power off.

**Comment by maehw** [comment]
Thank you for your detailed response. I may answer on snippets separately.

>Yes, it starts with the software. The web server is accessible on port 9080 on all interfaces.
>Just open a browser to http://localhost:9080

Note: Using only http://localhost:9080 will print out "Path not found".

Using http://localhost:9080/app (as seen in documentation before) automatically forwards to http://localhost:9080/app/index.html which shows a blank page for me. `view-source:http://localhost:9080/app/index.html` (Ctrl+U in Firefox) also is empty.

Edit: Parts of your answer have been modified and copied to the wiki.

**Reply by ClemensElflein** [reply]
I see, it might be that we have redirected stuff using NGINX. A blank page is expected if the agreement is not yet accepted. Maybe that's your issue here? Are there any errors in the browser log?

**Reply by maehw** [reply]
Good question.

As I was not aware of the webserver/web app I manually created `<HOME>/.selfomat_settings.json` as I saw it in the log files that such a file was needed. This is also why I manually set `show_agreement` to `false` without knowing about any cross-dependencies (related: #5):

```
{
    "show_agreement": "false",
    "trigger_counter": "102",
    "storage_enabled": "true",
    "printer_enabled": "false",
    "template_enabled": "false",
    "flash_enabled": "false",
    "filter_gain": "1",
    "filter_choice": "1",
    "debug_log_enabled": "false",
    "autofocus_before_trigger": "false",
    "frontend_language_choice": "1"
}
```

Could you give a hint where the NGINX configuration files are located within the file system?

**Reply by ClemensElflein** [reply]
I have mounted the file system and there is no magic in NGINX.. It just redirects from http://localhost/ to http://localhost:9080/app (without the index.html). So this should work normally without any issues. I have used it during development on my PC without any NGINX config.

The NGINX config file is located in in /etc/nginx/sites-available/default

**Reply by ClemensElflein** [reply]
Here are the file contents:
```
#
# You should look at the following URL's in order to grasp a solid understanding
# of Nginx configuration files in order to fully unleash the power of Nginx.
# https://www.nginx.com/resources/wiki/start/
# https://www.nginx.com/resources/wiki/start/topics/tutorials/config_pitfalls/
# https://wiki.debian.org/Nginx/DirectoryStructure
#
# In most cases, administrators will remove this file from sites-enabled/ and
# leave it as reference inside of sites-available where it will continue to be
# updated by the nginx packaging team.
#
# This file will automatically load configuration files provided by other
# applications, such as Drupal or Wordpress. These applications will be made
# available underneath a path with that package name, such as /drupal8.
#
# Please see /usr/share/doc/nginx-doc/examples/ for more detailed examples.
##

# Default server configuration
#
server {
        listen 80 default_server;
        listen [::]:80 default_server;

        # SSL configuration
        #
         listen 443 ssl default_server;
         listen [::]:443 ssl default_server;
        #
        # Note: You should disable gzip for SSL traffic.
        # See: https://bugs.debian.org/773332
        #
        # Read up on ssl_ciphers to ensure a secure configuration.
        # See: https://bugs.debian.org/765782
        #
        # Self signed certs generated by the ssl-cert package
        # Don't use them in a production server!
        #
         include snippets/snakeoil.conf;

        #root /var/www/html;

        # Add index.php to the list if you are using PHP
        #index index.html index.htm index.nginx-debian.html;

        server_name _;

        #location / {
                # First attempt to serve request as file, then
                # as directory, then fall back to displaying a 404.
        #       try_files $uri $uri/ =404;
        #}

        # pass PHP scripts to FastCGI server
        #
        #location ~ \.php$ {
        #       include snippets/fastcgi-php.conf;
        #
        #       # With php-fpm (or other unix sockets):
        #       fastcgi_pass unix:/run/php/php7.3-fpm.sock;
        #       # With php-cgi (or other tcp sockets):
        #       fastcgi_pass 127.0.0.1:9000;
        #}

        # deny access to .htaccess files, if Apache's document root
        # concurs with nginx's one
        #
        #location ~ /\.ht {
        #       deny all;
        #}
        return 302 http://$host:9080/app;
}



# Virtual Host configuration for example.com
#
# You can move that to a different file under sites-available/ and symlink that
# to sites-enabled/ to enable it.
#
#server {
#       listen 80;
#       listen [::]:80;
#
#       server_name example.com;
#
#       root /var/www/example.com;
#       index index.html;
#
#       location / {
#               try_files $uri $uri/ =404;
#       }
#}
```

**Reply by maehw** [reply]
Thank you for the update. Do you need to install/start NGINX explicitely? Or could it even be required to install something first?
(I am not really a web developer. I've only worked locally with Apache by running XAMPP back in those days or worked with preconfigured webservers.)

**Reply by maehw** [reply]
I am still having trouble getting this to work and could need some assistance here debugging the web server.

`GET http://localhost:9080/` returns 404,
`GET http://localhost:9080/app/index.html` returns 200 but no content
`GET http://localhost:9080/doesnotexist.html` also returns 200 and no content (I'd expect another 404)

Edit: FYI I am also working locally and not on a pre-installed Raspberry Pi.

**Reply by ClemensElflein** [reply]
Hi @maehw 
I think the web server should work out of the box without any NGINX running.
I'm setting up a dev system and try to get it to work. At the moment I cannot tell you why it's not working. Sorry!

I'll message you as soon as I know more.

**Reply by ClemensElflein** [reply]
As long as we haven't fixed the startup file, you can enable web server by copying the web/www folder to ./app.

The command would be ``cp -r web/www app`` (executed in repository root). Then the web app should be accessible without a problem.

**Reply by maehw** [reply]
If I get it right copying the files manually is no longer necessary using 514ec386220b5b3c83baf1d273e34f187b86151a.
http://localhost:9080/ still gives me "Path not found" - but if I understand correctly, this is due to missing NGINX forwarding configuration.
Using http://localhost:9080/app/ directly works and the web UI is shown, well done!  :partying_face:
Thanks for the fix!

**Reply by maehw** [reply]
For documentation purposes:

I tried to install and run `nginx` on my Raspberry Pi running the self-o-mat software so that I can access and debug it more easily from my developer's PC.

* Install `nginx`
  `$ sudo apt install nginx`
* Edit the configuration file and add the minimal configuration for proxying requests:
   ```
   # Minimal server configuration
   #
   server {
      listen 8080 default_server;   
      listen [::]:8080 default_server;

      server_name _;

      location / {
         proxy_pass http://127.0.0.1:9080;
      }
   }
   ```
* Reload
   `$ sudo nginx -s reload`
* Open web browser on development PC and access `http://<self-o-mat's IP address>:8080/app`
* Enjoy!

Disclaimer: The configuration might be working - but no guarantee that this is safe regarding IT security.

**Comment by ClemensElflein** [comment, marked as answer]
Fixed in 514ec386220b5b3c83baf1d273e34f187b86151a
